crates/environ/src/obj.rs

```diff
@@ -31,3 +32,67 @@ pub fn try_parse_trampoline_name(name: &str) -> Option<SignatureIndex> {
     let n = name.strip_prefix(TRAMPOLINE_PREFIX)?.parse().ok()?;
     Some(SignatureIndex::new(n))
 }
+
+/// Returns an iterator of the wasm function index and corresponding symbol for
+/// the function given a module/compilation artifacts.
+///
+/// This function only works with Wasmtime's compilation artifacts and isn't
+/// supposed to be general purpose otherwise. The shape of the `obj` passed in
+/// is implicitly assumed and runtime checks will fail if the `obj` is
+/// malformed.
+pub fn defined_functions<'a, 'data>(
+    module: &'a Module,
+    obj: &'a File<'data>,
+) -> impl Iterator<Item = (DefinedFuncIndex, Symbol<'data, 'a>)> + 'a {
+    // The goal here is to determine the `SymbolIndex` for a `DefinedFuncIndex`
+    // without actually synthesizing the name of the string or doing searching
+    // in the symbol table otherwise.
+    //
+    // Given how we construct object files we are guaranteed that all local
+    // symbols are defined first, and of those local symbols the first symbols
+    // are the defined functions of the module, in sorted order of index. This
+    // means that there's a pretty easy mapping from `DefinedFuncIndex` to
+    // `SymbolIndex`.
+    //
+    // Note, though, that we skip the first entry, the 0th symbol, since the
+    // `object` crate reserves that for a null symbol of some sort.
```

Comment:
This mapping feels pretty brittle, and likely to break accidentally with a future change. Is there a way that we could export a helper from the `object` crate (or factor into a third crate if that's an unwanted dep) that turns this pseudo-specified mapping into a more solid thing?

Reply:
(On reading further down, I do see the debug-assert check of this invariant and the related one for trampoline functions, *but* it still feels somewhat fragile to me as a sort of distributed invariant.)